server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -116,55 +119,71 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
-                            exception);
-                    }
-                }
+            final IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            if (indexShard != null) {
+                final ShardStateMetaData shardStateMetaData = indexShard.getShardStateMetaData();
+                final String allocationId = shardStateMetaData.allocationId != null ?
+                    shardStateMetaData.allocationId.getId() : null;
+                logger.trace("{} shard state info found: [{}]", shardId, shardStateMetaData);
+                return new NodeGatewayStartedShards(clusterService.localNode(), allocationId, shardStateMetaData.primary);
+            }
+
+            final ShardStateMetaData shardStateMetaData;
+            try (ShardLock ignored = nodeEnv.shardLock(shardId, TimeUnit.SECONDS.toMillis(5))) {
+                shardStateMetaData = ShardStateMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY,
+                    nodeEnv.availableShardPaths(shardId));
+            }
+
+            if (shardStateMetaData == null) {
+                logger.trace("{} no local shard info found", shardId);
+                return new NodeGatewayStartedShards(clusterService.localNode(), null, false);
+            }
 
-                logger.debug("{} shard state info found: [{}]", shardId, shardStateMetaData);
+            IndexMetaData metaData = clusterService.state().metaData().index(shardId.getIndex());
+            if (metaData == null) {
+                // we may send this requests while processing the cluster state that recovered the index
+                // sometimes the request comes in before the local node processed that cluster state
+                // in such cases we can load it from disk
+                metaData = IndexMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY,
+                    nodeEnv.indexPaths(shardId.getIndex()));
+            }
+            if (metaData == null) {
+                ElasticsearchException e = new ElasticsearchException("failed to find local IndexMetaData");
+                e.setShard(request.shardId);
+                throw e;
+            }
+
+            // we don't have an open shard on the store, validate the files on disk are openable
+            ShardPath shardPath = null;
+            try {
+                IndexSettings indexSettings = new IndexSettings(metaData, settings);
+                try (ShardLock ignored = nodeEnv.shardLock(shardId, TimeUnit.SECONDS.toMillis(5))) {
+                    shardPath = ShardPath.loadShardPath(logger, nodeEnv, shardId, indexSettings);
+                }
+                if (shardPath == null) {
+                    throw new IllegalStateException(shardId + " no shard path found");
+                }
+                Store.tryOpenIndex(shardPath.resolveIndex(), shardId, nodeEnv::shardLock, logger);
```

Comment:
Instead of acquiring the shard lock for a second time, I would prefer if we would do it once, and move this call under that lock and just rename `tryOpenIndex` to `tryOpenIndexUnderLock`, removing the locking mechanism from it.

Same thing for `TransportNodesListShardStoreMetaData`. You can then also remove the `ShardLocker` interface, which irked me for a while.